File: advanced_localization_cpp/src/math_utils.cpp
#include "advanced_localization_cpp/math_utils.hpp"

#include <cmath>

namespace advanced_localization
{

Matrix3 skew_symmetric(const Vector3 & vector)
{
  Matrix3 matrix;
  matrix << 0.0, -vector.z(), vector.y(),
    vector.z(), 0.0, -vector.x(),
    -vector.y(), vector.x(), 0.0;
  return matrix;
}

double normalize_angle(double angle_rad)
{
  return std::atan2(std::sin(angle_rad), std::cos(angle_rad));
}

Eigen::VectorXd clip_vector_norm(const Eigen::VectorXd & vector, double max_norm)
{
  const double norm = vector.norm();
  if (norm <= max_norm || norm < 1e-12) {
    return vector;
  }
  return vector * (max_norm / norm);
}

Vector4 quaternion_normalize(const Vector4 & quaternion_wxyz)
{
  const double norm = quaternion_wxyz.norm();
  if (norm < 1e-12) {
    return Vector4(1.0, 0.0, 0.0, 0.0);
  }
  return quaternion_wxyz / norm;
}

Vector4 quaternion_multiply(const Vector4 & lhs_wxyz, const Vector4 & rhs_wxyz)
{
  const auto lhs = quaternion_normalize(lhs_wxyz);
  const auto rhs = quaternion_normalize(rhs_wxyz);
  const double lw = lhs(0), lx = lhs(1), ly = lhs(2), lz = lhs(3);
  const double rw = rhs(0), rx = rhs(1), ry = rhs(2), rz = rhs(3);
  return Vector4(
    lw * rw - lx * rx - ly * ry - lz * rz,
    lw * rx + lx * rw + ly * rz - lz * ry,
    lw * ry - lx * rz + ly * rw + lz * rx,
    lw * rz + lx * ry - ly * rx + lz * rw);
}

Vector4 quaternion_from_small_angle(const Vector3 & delta_theta_rad)
{
  const double theta = delta_theta_rad.norm();
  if (theta < 1e-12) {
    return quaternion_normalize(Vector4(1.0, 0.5 * delta_theta_rad.x(), 0.5 * delta_theta_rad.y(), 0.5 * delta_theta_rad.z()));
  }
  const double half_theta = 0.5 * theta;
  const Vector3 axis = delta_theta_rad / theta;
  const double sin_half = std::sin(half_theta);
  return quaternion_normalize(Vector4(std::cos(half_theta), axis.x() * sin_half, axis.y() * sin_half, axis.z() * sin_half));
}

Matrix3 rotation_matrix_from_quaternion(const Vector4 & quaternion_wxyz)
{
  const auto q = quaternion_normalize(quaternion_wxyz);
  const double w = q(0), x = q(1), y = q(2), z = q(3);
  Matrix3 matrix;
  matrix <<
    1.0 - 2.0 * (y * y + z * z), 2.0 * (x * y - w * z), 2.0 * (x * z + w * y),
    2.0 * (x * y + w * z), 1.0 - 2.0 * (x * x + z * z), 2.0 * (y * z - w * x),
    2.0 * (x * z - w * y), 2.0 * (y * z + w * x), 1.0 - 2.0 * (x * x + y * y);
  return matrix;
}

double yaw_from_quaternion(const Vector4 & quaternion_wxyz)
{
  const auto q = quaternion_normalize(quaternion_wxyz);
  const double w = q(0), x = q(1), y = q(2), z = q(3);
  return normalize_angle(std::atan2(2.0 * (w * z + x * y), 1.0 - 2.0 * (y * y + z * z)));
}

Vector4 quaternion_from_yaw(double yaw_rad)
{
  const double half = 0.5 * yaw_rad;
  return Vector4(std::cos(half), 0.0, 0.0, std::sin(half));
}

Matrix16 force_symmetric(const Matrix16 & matrix)
{
  return 0.5 * (matrix + matrix.transpose());
}

Eigen::MatrixXd force_symmetric_dynamic(const Eigen::MatrixXd & matrix)
{
  return 0.5 * (matrix + matrix.transpose());
}

Vector3 enu_to_ned_position(const Vector3 & enu)
{
  return Vector3(enu.y(), enu.x(), -enu.z());
}

Vector3 ned_to_enu_position(const Vector3 & ned)
{
  return Vector3(ned.y(), ned.x(), -ned.z());
}

Vector4 enu_quaternion_to_ned(const Vector4 & q_enu_wxyz)
{
  constexpr double half_pi = 1.57079632679489661923;
  const double yaw_ned = normalize_angle(half_pi - yaw_from_quaternion(q_enu_wxyz));
  return quaternion_from_yaw(yaw_ned);
}

}  // namespace advanced_localization

File: src/advanced_localization/math_utils.py
from __future__ import annotations

import numpy as np
from numpy.typing import NDArray


def skew_symmetric(vector: NDArray[np.float64]) -> NDArray[np.float64]:
    x, y, z = vector
    return np.array(
        [
            [0.0, -z, y],
            [z, 0.0, -x],
            [-y, x, 0.0],
        ],
        dtype=np.float64,
    )


def normalize_angle(angle_rad: float) -> float:
    return float(np.arctan2(np.sin(angle_rad), np.cos(angle_rad)))


def clip_vector_norm(vector: NDArray[np.float64], max_norm: float) -> NDArray[np.float64]:
    norm = float(np.linalg.norm(vector))
    if norm <= max_norm or norm < 1e-12:
        return vector
    return vector * (max_norm / norm)


def quaternion_normalize(quaternion_wxyz: NDArray[np.float64]) -> NDArray[np.float64]:
    norm = float(np.linalg.norm(quaternion_wxyz))
    if norm < 1e-12:
        return np.array([1.0, 0.0, 0.0, 0.0], dtype=np.float64)
    return quaternion_wxyz / norm


def quaternion_multiply(lhs_wxyz: NDArray[np.float64], rhs_wxyz: NDArray[np.float64]) -> NDArray[np.float64]:
    lw, lx, ly, lz = lhs_wxyz
    rw, rx, ry, rz = rhs_wxyz
    return np.array(
        [
            lw * rw - lx * rx - ly * ry - lz * rz,
            lw * rx + lx * rw + ly * rz - lz * ry,
            lw * ry - lx * rz + ly * rw + lz * rx,
            lw * rz + lx * ry - ly * rx + lz * rw,
        ],
        dtype=np.float64,
    )


def quaternion_from_small_angle(delta_theta_rad: NDArray[np.float64]) -> NDArray[np.float64]:
    theta = float(np.linalg.norm(delta_theta_rad))
    if theta < 1e-12:
        return quaternion_normalize(
            np.array([1.0, 0.5 * delta_theta_rad[0], 0.5 * delta_theta_rad[1], 0.5 * delta_theta_rad[2]], dtype=np.float64)
        )

    half_theta = 0.5 * theta
    axis = delta_theta_rad / theta
    sin_half = np.sin(half_theta)
    return quaternion_normalize(
        np.array(
            [
                np.cos(half_theta),
                axis[0] * sin_half,
                axis[1] * sin_half,
                axis[2] * sin_half,
            ],
            dtype=np.float64,
        )
    )


def rotation_matrix_from_quaternion(quaternion_wxyz: NDArray[np.float64]) -> NDArray[np.float64]:
    w, x, y, z = quaternion_normalize(quaternion_wxyz)
    xx = x * x
    yy = y * y
    zz = z * z
    xy = x * y
    xz = x * z
    yz = y * z
    wx = w * x
    wy = w * y
    wz = w * z

    return np.array(
        [
            [1.0 - 2.0 * (yy + zz), 2.0 * (xy - wz), 2.0 * (xz + wy)],
            [2.0 * (xy + wz), 1.0 - 2.0 * (xx + zz), 2.0 * (yz - wx)],
            [2.0 * (xz - wy), 2.0 * (yz + wx), 1.0 - 2.0 * (xx + yy)],
        ],
        dtype=np.float64,
    )


def yaw_from_quaternion(quaternion_wxyz: NDArray[np.float64]) -> float:
    w, x, y, z = quaternion_normalize(quaternion_wxyz)
    yaw = np.arctan2(2.0 * (w * z + x * y), 1.0 - 2.0 * (y * y + z * z))
    return normalize_angle(float(yaw))


def quaternion_from_yaw(yaw_rad: float) -> NDArray[np.float64]:
    half = 0.5 * yaw_rad
    return np.array([np.cos(half), 0.0, 0.0, np.sin(half)], dtype=np.float64)


def force_symmetric(matrix: NDArray[np.float64]) -> NDArray[np.float64]:
    return 0.5 * (matrix + matrix.T)
